### ISSUE TITLE
Request for adding implementations of scipy hypergeometric functions

### ISSUE THREAD

**Issue by shiningalehxb** [cupy/cupy#8274]
### Description

Hello cupy develop team,

I would like to request whether the scipy features of evaluating hypergeometric functions, such as scipy.special.hyper2f1, scipy.special.hyper1f1, etc., can be incorporated into cupy. I have browsed the online documentation of cupyx.scipy.special but did not find hypergeometric functions in the list of special functions. The reason for such feature request is that, I need to frequently generate multi-dimensional Monte Carlo simulation paths where Gauss hypergeometric functions are heavily needed in constructing covariance matrices for random vectors. Thank you! :)

**Comment by steppi** [comment]
Hi @shiningalehxb. In SciPy, we've recently rewritten the complex valued `hyp2f1` in C++ which can run on CUDA, https://github.com/scipy/scipy/pull/20089. The real valued `hyp2f1` will available when https://github.com/scipy/scipy/pull/20390 is merged. I anticipate being able to submit a PR to CuPy adding `hyp2f1` within the next few weeks. Rewriting `hyp1f1` in such a way is on our horizon, and will likely be finished in the next few months.

My collaborator @izaid and I are working on a C++ library of special functions which can run on both CPU and GPU that could be used by SciPy, CuPy, and other array libraries. See https://github.com/cupy/cupy/pull/8140 for an example of a special function we've already added in this way. We hope for this library to encompass all or most of what's now in `scipy.special`.

**Comment by shiningalehxb** [comment]
> Hi @shiningalehxb. In SciPy, we've recently rewritten the complex valued `hyp2f1` in C++ which can run on CUDA, [scipy/scipy#20089](https://github.com/scipy/scipy/pull/20089). The real valued `hyp2f1` will available when [scipy/scipy#20390](https://github.com/scipy/scipy/pull/20390) is merged. I anticipate being able to submit a PR to CuPy adding `hyp2f1` within the next few weeks. Rewriting `hyp1f1` in such a way is on our horizon, and will likely be finished in the next few months.
> 
> My collaborator @izaid and I are working on a C++ library of special functions which can run on both CPU and GPU that could be used by SciPy, CuPy, and other array libraries. See #8140 for an example of a special function we've already added in this way. We hope for this library to encompass all or most of what's now in `scipy.special`.

Thank you very much for your kind and helpful reply :)

**Comment by dhruvdcoder** [comment]
It would be nice to have `hyp1f1` in cupy as well.

**Comment by maciejskorski** [comment]
It would be very helpful to benefit of GPU-accelerated hypergeometric functions.

The referenced SciPy PR has been merged already, what is the blocker now? What (if any) help is needed?

**Comment by steppi** [comment]
Hi @maciejskorski. I've submitted a PR here to add https://github.com/cupy/cupy/pull/9159 the `xsf` library that's now a submodule of SciPy as a submodule to CuPy. Unfortunately, https://github.com/scipy/scipy/pull/20390, didn't actually unblock having a real valued `hyp2f1` implementation because SciPy's real valued implementation relies on recursion to handle recurrences for argument reduction, which is not viable with CUDA, and the structure of the code made it challenging to bypass the recursion given the bandwidth I had available. Instead I plan to make the complex implementation I wrote generic and viable for real arguments; though it won't have argument reduction for large $|a|$, $|b|$, $|c|$ and thus will be inaccurate in these regimes. I would then make `xsf` use conditional compilation to choose a different implementation depending on whether one is targeting CPU or GPU.

I think what needs to happen to get a GPU accelerated `hyp2f1` is

1. https://github.com/scipy/xsf/pull/39 lands in the `xsf` repo. This is a blocker for https://github.com/cupy/cupy/pull/9159 . I'm waiting on a review from another SciPy maintainer for this one.
2. https://github.com/cupy/cupy/pull/9159 lands in CuPy.
3. An option is added to set the language standard for ufuncs created in CuPy through [cupy._core.create_ufunc](https://github.com/cupy/cupy/blob/fc5ee663380ff211eadbf7890096f2d46c639922/cupy/_core/_kernel.pyx#L1691). `xsf` is a C++17 codebase, and some C++17 features are to get everything working without considerable extra effort.
4. I or someone else does the work to make a real valued `hyp2f1` suitable for GPU and lands a PR in `xsf`. This requires someone to do the work, and someone to review the PR. There could then be a follow-up to update the `xsf` submodule in CuPy and define a `hyp2f1` CuPy `ufunc`.

Getting reviews to PRs I make to SciPy / `xsf` has been a substantial bottleneck; I'm not sure if you have bandwidth, but it would make a big impact if you could help with that.
